### ISSUE TITLE
how to disable push for a project

### ISSUE THREAD

**Issue by roldancer** [vmware/harbor#2331]
Hi, Is it possible to disable push for a project using the admin user ?

As a Harbor administrator I would like to disable push command for a project, some users could push lot of images to the repository so I need to prevent filesystem exhaustion, to prevent this kind of DOS I have to disable the push operation only for the projects that are consuming filesystem free space, it is possible ?

By the way, the project's admin should not enable push operation if the Harbor's admin have disabled it.
Thanks,

**Comment by reasonerjt** [comment]
So you want the system admin to disable push for a particular project? or set the whole Harbor to read-only?

Technically this is possible, but from a permission control point of view, the action targeting to a specific project, should be controlled by project admin instead of system admin.  Otherwise it may cause confusion...

**Comment by roldancer** [comment]
Hi @reasonerjt 

     I would like system admin to disable some specific projects, not all the registry.

**Comment by reasonerjt** [comment]
@roldancer thanks for clarification, we need to think about it in the future, but I personally have concerns about this design in terms of user experience and consistency of permission model.

**Comment by roldancer** [comment]
@reasonerjt sure we have to think about it, but at the moment any user with push capability could generate  a DOS issue (filesystem exhaustion), the only way to minimize this kind of DOS would be disabling the push operation for the project that is generating the filesystem exhaustion.

**Comment by reasonerjt** [comment]
@roldancer 
1) not any user, only the ones that project admin grant permission to.
2) If you set some of the projects to "readonly" it doesn't prevent DOS, other projects are still under risk.

**Comment by roldancer** [comment]
@reasonerjt from my point of view this feature (disabling/enabling push operation) is the first step for project quota/limits management.

**Comment by reasonerjt** [comment]
OK, so that makes sense, maybe we should skip the "readonly" and design the quote/limits feature.


What should be controlled in your opinion in regards to quota/limits?  Given the fact that docker images share layers, controlling the storage under certain project doesn't make sense.  Maybe control the number of images, or the number of tags of a repo?

**Comment by roldancer** [comment]
@reasonerjt I agree with you controlling the number of images or tags should be the right metric, the other question is what we should do if a project reach the limit of images/tag ? from my point of view we should put that project in "readonly" mode.

**Comment by akshayl** [comment]
@roldancer In your Harbor setup are users allowed to create projects or is that disabled?

**Comment by roldancer** [comment]
@akshayl project creation is disabled, only Harbor's administrator can generates projects.

**Comment by gunboe** [comment]
A quota limit is important to me. In my environment, users consume a lot of storage, many projects are created without "the images best practice", so big images are made. Therefore, I need to establish storage quotas per user or per project.

**Comment by reasonerjt** [comment]
This is essentially a requirement of quota based on size, but so far we don't have a good way to figure out how much space a project actually uses.

I'll put this to backlog.

**Comment by tianon** [comment]
Project Quotas was implemented in 1.9 (https://github.com/goharbor/harbor/issues/1005, esp https://github.com/goharbor/harbor/issues/1005#issuecomment-541124850), but IMO there's still a good use case for being able to make the entire registry or just a project "read-only" similar to GitHub's archived repositories.

As a Harbor system administrator, I may be migrating to another repository but still need users and my new repository to be able to access the existing images for historical purposes (which is a feature that exists today in the System Settings as "Repository Read Only").

As a project manager, I might have the same thing but on a smaller scale -- perhaps I was working on a side project within the organization that is no longer supported, so making the project read-only would be a useful way to signal that there should not be any further changes (the best workaround I can come up with for accomplishing this today is to downgrade all members to a role that doesn't allow them to push, but that doesn't help prevent an administrator from accidentally modifying the archived project).
